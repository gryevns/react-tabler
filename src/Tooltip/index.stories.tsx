import React from "react";
import { Heart } from "react-feather";

import Buttons from "../Buttons";
import Card from "../Card";
import Tooltip from "../Tooltip";

export default {
    title: "Tooltip",
};

const placements = [
    "top" as const,
    "right" as const,
    "bottom" as const,
    "left" as const,
];

export const Placement = () => (
    <Card size="lg">
        <Card.Body>
            <Buttons.List>
                {placements.map((placement) => (
                    <Tooltip content="Tooltip Content" placement={placement}>
                        <Buttons.Button variant="secondary" className="mb-0">
                            Tooltip on {placement}
                        </Buttons.Button>
                    </Tooltip>
                ))}
            </Buttons.List>
        </Card.Body>
    </Card>
);

export const Elements = () => (
    <Tooltip content="Tooltip Content" placement="right">
        <div style={{ display: "inline-block", margin: 10 }}>
            <Heart size={18} />
        </div>
    </Tooltip>
);

export const Paragraph = () => (
    <Card>
        <Card.Body>
            <p>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam
                ullamcorper porta viverra.{" "}
                <Tooltip content="Tooltip Content" placement="bottom">
                    <u>
                        <b>Class aptent taciti</b>
                    </u>
                </Tooltip>{" "}
                sociosqu ad litora torquent per conubia nostra, per inceptos
                himenaeos. Donec vel gravida lacus, vel vestibulum sem. Donec blandit
                elit id sem suscipit aliquam. Praesent tellus lectus, bibendum ac
                convallis et, cursus at dolor. Vivamus enim arcu, rutrum quis ex
                eget,{" "}
                <Tooltip content="Tooltip Content" placement="top">
                    <u>
                        <b>venenatis fermentum dolor</b>
                    </u>
                </Tooltip>
                . Maecenas feugiat leo quis orci aliquet, in accumsan neque dapibus.
                Suspendisse magna velit, pulvinar id accumsan nec, dapibus nec
                mauris. Quisque finibus erat pharetra sollicitudin condimentum.{" "}
            </p>
        </Card.Body>
    </Card>
);
