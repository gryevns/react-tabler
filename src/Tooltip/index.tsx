import React from "react";
import cn from "classnames";
import { Manager, Reference, Popper, PopperArrowProps } from "react-popper";

const defaultModifiers = [
    {
        name: "offset",
        options: {
            offset: [0, 4],
        },
    },
];

const Arrow = React.forwardRef(
    (props: PopperArrowProps, ref: React.Ref<HTMLDivElement>) => (
        <div className="tooltip-arrow" ref={ref} data-popper-arrow {...props} />
    ),
);

export interface TooltipProps {
    children: React.ReactElement;
    placement: "top" | "right" | "bottom" | "left";
    className?: string;
    arrow?: boolean;
    content: React.ReactNode;
    modifiers?: any[];
}

const Tooltip = ({
    children,
    placement = "right",
    className,
    arrow = true,
    modifiers = defaultModifiers,
    content,
}: TooltipProps) => {
    const [visible, toggle] = React.useReducer((state) => !state, false);

    const classes = cn(
        "tooltip",
        placement && "bs-tooltip-" + placement,
        "show",
        className,
    );

    return (
        <Manager>
            <Reference>
                {({ ref }) =>
                    React.cloneElement(children, {
                        ref,
                        onMouseOver: toggle,
                        onMouseOut: toggle,
                    })
                }
            </Reference>
            {visible ? (
                <Popper placement={placement} modifiers={modifiers}>
                    {({ ref, style, placement, arrowProps }) => (
                        <div
                            className={classes}
                            data-placement={placement}
                            style={style}
                            ref={ref}
                        >
                            {arrow && <Arrow {...arrowProps} />}
                            <div className="tooltip-inner">{content}</div>
                        </div>
                    )}
                </Popper>
            ) : null}
        </Manager>
    );
};

export default Tooltip;
