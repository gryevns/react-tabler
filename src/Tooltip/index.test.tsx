import React from "react";
import { shallow } from "enzyme";

import Tooltip from "./index";

describe("Tooltip", () => {
    it("should match toast snapshot", () => {
        const wrapper = shallow(
            <Tooltip placement="bottom" content="test">
                <div />
            </Tooltip>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
