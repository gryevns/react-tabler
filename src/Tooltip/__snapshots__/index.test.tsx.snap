// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match toast snapshot 1`] = `
<Manager>
  <Reference>
    <Component />
  </Reference>
</Manager>
`;
